### PR TITLE
8264337: VM crashed when -XX:+VerifySharedSpaces

### PR DESCRIPTION
**Pull request by kelthuzadx**
`java -XX:+VerifySharedSpaces -version` reports an internal error that would crash VM:

```
[0.113s][error][cds] relocation bitmap CRC error
[0.114s][error][cds] relocation bitmap CRC error
# To suppress the following error report, specify this argument
# after -XX: or in .hotspotrc: SuppressErrorAt=/javaClasses.inline.hpp:215
#
# A fatal error has been detected by the Java Runtime Environment:
#
# Internal Error (/home/qingfeng.yy/openjdk16_so_warning/jdk/src/hotspot/share/classfile/javaClasses.inline.hpp:215), pid=75241, tid=75242
# assert(java_lang_Class::is_instance(java_class)) failed: must be a Class object
#
# JRE version: (17.0) (slowdebug build )
# Java VM: OpenJDK 64-Bit Server VM (slowdebug 17-internal+0-adhoc.qingfengyy.jdk, mixed mode, sharing, tiered, compressed oops, compressed class ptrs, g1 gc, linux-amd64)
# Problematic frame:
# V [libjvm.so+0x697d19] java_lang_Class::as_Klass(oopDesc*)+0x29
#
# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /home/qingfeng.yy/openjdk16_so_warning/jdk/build/linux-x86_64-server-slowdebug/images/jdk/bin/hs_err_pid75241.log
#
#
```
The gut of this problem is that the calculation of crc32 for the CDS archive is inconsistent between dumptime and runtime, this will cause subsequent steps to deviate from the desired path one after another, eventually leading to an internal error that crashes the VM. 

All tests under hotspot/jtreg/runtime/cds are passed(excepts those problem-listed) with slowdebug mode.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8264337](https://bugs.openjdk.java.net/browse/JDK-8264337): VM crashed when -XX:+VerifySharedSpaces

### Reviewers
 * [Yumin Qi](https://openjdk.java.net/census#minqi) (@yminqi - **Reviewer**) ⚠️ Review applies to 64780bc16110a6ddf2a30a1b9666213d5cf333d0
 * [Calvin Cheung](https://openjdk.java.net/census#ccheung) (@calvinccheung - **Reviewer**) ⚠️ Review applies to 64780bc16110a6ddf2a30a1b9666213d5cf333d0
 * [Aleksey Shipilev](https://openjdk.java.net/census#shade) (@shipilev - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3244/head:pull/3244` \
`$ git checkout pull/3244`

Update a local copy of the PR: \
`$ git checkout pull/3244` \
`$ git pull https://git.openjdk.java.net/jdk pull/3244/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3244`

View PR using the GUI difftool: \
`$ git pr show -t 3244`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3244.diff">https://git.openjdk.java.net/jdk/pull/3244.diff</a>

</details>

### CODE REVIEW

**Review by yminqi — APPROVED** [review]
You are right. The crc is calculated based on used size, not aligned used size.

**Review by calvinccheung — APPROVED** [review]
LGTM.

**Comment by kelthuzadx** [comment]
Thanks @yminqi @calvinccheung for reviews！

/integrate

File: test/hotspot/jtreg/runtime/cds/appcds/VerifyWithDefaultArchive.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import jdk.test.lib.process.OutputAnalyzer;
+
+public class VerifyWithDefaultArchive {
+    public static void main(String... args) throws Exception {
+        ProcessBuilder pb = ProcessTools.createJavaProcessBuilder("-Xlog:cds", "-XX:+VerifySharedSpaces", "-version");
+        OutputAnalyzer out = new OutputAnalyzer(pb.start());
+        out.shouldContain("OpenJDK");
+        out.shouldHaveExitValue(0);
+    }
+}
```
Comment (shipilev) [inline test/hotspot/jtreg/runtime/cds/appcds/VerifyWithDefaultArchive.java:43]: Please add the newline at the end of file here. See how UI complains about it :)

**Review by shipilev — APPROVED** [review]
Minor nit.

File: test/hotspot/jtreg/runtime/cds/appcds/VerifyWithDefaultArchive.java (hunk 1 above)
Comment (kelthuzadx) [reply test/hotspot/jtreg/runtime/cds/appcds/VerifyWithDefaultArchive.java:43]: Got it. But I'm not sure what's the abbreviation `UI` means... Could you please give me a link for that? Thanks!

Comment (shipilev) [reply test/hotspot/jtreg/runtime/cds/appcds/VerifyWithDefaultArchive.java]: Thanks! I meant how GitHub UI (user interface) shows the red stop sign in cases like these.

**Review by shipilev — APPROVED** [review]

File: test/hotspot/jtreg/runtime/cds/appcds/VerifyWithDefaultArchive.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+/*
+ * @test 8264337
```
Comment (shipilev) [inline test/hotspot/jtreg/runtime/cds/appcds/VerifyWithDefaultArchive.java:26]: Sorry, one other minor nit. This should be:
```
@test
@bug 8264337
```

**Review by shipilev — APPROVED** [review]
Okay, we should be good to go. Thanks!

**Comment by kelthuzadx** [comment]
Thank you @shipilev, let me
/integrate
it again!

**Comment by shipilev** [comment]
I also ran `runtime/cds` tests without problem.

/sponsor

**Pull request closed without merging**